### PR TITLE
Changing 1-to-M behaviour of on_disk_cache. (#810)

### PR DESCRIPTION
**Pull request by ejguan**
Summary:
Pull Request resolved: https://github.com/pytorch/data/pull/810

TLDR: If filename_fn of on_disk_cache ( filename_fn'1 ) generates different name from filename_fn of end_caching, it is considered 1 to many situation. In this case additional listing file would be created at filename_fn'1 position and it will be used to check cache consistency between runs.

BC breaking, on_disk_cache no longer accept generators as `filename_fn`

Test Plan: Imported from OSS

Reviewed By: ejguan

Differential Revision: D40148560

Pulled By: VitalyFedyunin

fbshipit-source-id: 3bb9c5f32546a3d4859cff6bda0cc84234312ab7

Please read through our [contribution guide](https://github.com/pytorch/data/blob/main/CONTRIBUTING.md) prior to
creating your pull request.

- Note that there is a section on requirements related to adding a new DataPipe.

Fixes #{issue number}

### Changes

-
-

### CODE REVIEW

**Comment by ejguan** [comment]
I need to merge it first before TorchText to update RC release

**Review by NivekT — APPROVED** [review]
LGTM!
